[issue-3535] [FE] Make MetricDateRangeSelect interval calculation UTC-deterministic (fix local TZ test flake)

yashx007 opened the pull request:
# [OPIK-3535] Make date comparison UTC-aware and add CI timezone test matrix

## Description

This PR fixes a timezone-dependent test flake in the frontend by making the day comparison deterministic using UTC. It also adds a lightweight GitHub Actions workflow that runs frontend unit tests in multiple timezones to catch similar regressions early in CI.


## What changed

* **utils.ts**

  * Updated day comparison to use UTC normalization when deciding if the "to" date is "today".
  * Preserved the injectable `currentTime` parameter (used in tests) to minimize behavior changes, only removing timezone sensitivity.

* **utils.test.ts**

  * Adjusted tests to use a fixed `mockCurrentDate` and rely on deterministic UTC comparisons.

* **frontend-tests-timezone.yml**

  * Added a new CI workflow running frontend tests with environment variable `TZ` set to:

    * `UTC`
    * `America/Los_Angeles`
  * Purpose: to catch timezone-sensitive test regressions early.


## Why this change

Tests were flaky depending on the developer machine timezone because `dayjs().isSame(..., "day")` used local time implicitly. Normalizing both sides to UTC ensures consistent behavior regardless of the runner's timezone.

Adding the CI timezone matrix guards automatically against timezone-specific regressions.

---

## Behavior and safety

* No runtime behavior changes except that date comparisons are explicitly normalized to UTC to avoid ambiguous “today” semantics when comparing user-supplied dates vs server/now across timezones.
* Change is minimal and localized to the date-range utility and its tests.


## How to test locally

1. Navigate to the frontend app directory:

   ```bash
   cd apps/opik-frontend
   ```

2. Install dependencies (if needed):

   ```bash
   npm ci
   ```

3. Run the targeted tests:

   ```bash
   npx vitest run src/components/pages-shared/traces/MetricDateRangeSelect/utils.test.ts
   ```

4. To verify cross-timezone behavior locally (PowerShell example):

   ```powershell
   # Run in UTC
   $env:TZ='UTC'
   npx vitest run src/components/pages-shared/traces/MetricDateRangeSelect/utils.test.ts
   Remove-Item Env:\TZ

   # Run in America/Los_Angeles
   $env:TZ='America/Los_Angeles'
   npx vitest run src/components/pages-shared/traces/MetricDateRangeSelect/utils.test.ts
   Remove-Item Env:\TZ
   ```

5. (Optional) Run full frontend tests:

   ```bash
   npx vitest run
   ```


##Screenshots

<img width="1480" height="623" alt="Screenshot 2025-10-05 143646" src="https://github.com/user-attachments/assets/51494ad6-b821-4929-a36f-4afaa6110e7c" />

<img width="1477" height="620" alt="Screenshot 2025-10-05 143700" src="https://github.com/user-attachments/assets/20340f89-0ccf-497d-8785-a1627f14e3d5" />



## Notes for reviewers

* Focus on the UTC normalization in the `isEndDateToday` logic in `utils.ts`.
* Check the test assertions related to `intervalEnd` and `intervalStart`.
* The new workflow is small, running only frontend unit tests with a TZ matrix.
* If preferred, timezone matrix can be expanded (e.g., add `Asia/Shanghai` or `Europe/London`).



## Checklist before merge

* [ ] Code review of `utils.ts` UTC comparison logic
* [ ] Tests pass locally (targeted test file)
* [ ] Branch pushed including new workflow file

Fixes #3535

yashx007 commented:
@vincentkoc can u review it?

yashx007 commented on apps/opik-frontend/src/components/pages-shared/traces/MetricDateRangeSelect/utils.test.ts:
```diff
@@ -248,7 +248,7 @@ describe("calculateIntervalStartAndEnd", () => {
       };
 
-      const result = calculateIntervalStartAndEnd(dateRange);
+  const result = calculateIntervalStartAndEnd(dateRange, new Date(mockCurrentDate));
```
fixed

aadereiko commented on apps/opik-frontend/src/components/pages-shared/traces/MetricDateRangeSelect/utils.ts:
```diff
@@ -1,6 +1,9 @@
 import dayjs from "dayjs";
+import utc from "dayjs/plugin/utc";
+
+dayjs.extend(utc);
```
I don't think you should extend `utc` here as it should be done only once (and it's dont in  `opik/apps/opik-frontend/src/lib/date.ts`)

aadereiko commented on node_modules/.vite/vitest/da39a3ee5e6b4b0d3255bfef95601890afd80709/results.json:
Could you remove this file as well?

yashx007 commented:
hey @aadereiko just wait am doing. i will let u know once done

yashx007 commented:
Hi @aadereiko,
I've made all the changes you requested, removed the duplicate UTC extension and cleaned up the code. The results.json file was a stale tracked file and has now been removed from git. Thanks for your review!

Nimrod007 commented:
Hey @yashx007 - thanks for the contribution, im going to close this since we put this issue in low priority. (you can run the tests locally in docker)

feel free to continue to contribute in the future! and next time worth just fixing the minmal part and getting approval (vs adding github actions or some out of scope content)
